File: core/compiler/semantic_lowering_engine.py
from __future__ import annotations

from typing import Any, Dict, List


MAX_LOWERED_EDGES = 10000


def lower_semantic_ir(ir: Dict[str, Any]) -> Dict[str, Any]:
    edges: List[Dict[str, Any]] = list(ir.get("edges", []))[:MAX_LOWERED_EDGES]

    lowered = []

    for edge in edges:
        lowered.append({
            "source": edge.get("from"),
            "target": edge.get("to"),
            "relationship": edge.get("type", "semantic_link"),
        })

    return {
        "lowered_edges": lowered,
        "bounded": True,
        "count": len(lowered),
    }

File: src/compiler/semanticLoweringEngine.ts
/**
 * Converted from Python: core/compiler/semantic_lowering_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_LOWERED_EDGES: any = 10000;
export function lowerSemanticIr(ir: any): any {
  var edges: any = py.slice([...py.iter(py.get(ir, "edges", []))], null, MAX_LOWERED_EDGES);
  var lowered: any[] = [];
  var edge: any;
  for (edge of py.iter(edges)) {
    py.listAppend(lowered, {"source": py.get(edge, "from"), "target": py.get(edge, "to"), "relationship": py.get(edge, "type", "semantic_link")});
  }
  return {"lowered_edges": lowered, "bounded": true, "count": py.len(lowered)};
}
